### ISSUE TITLE
Shady DOM requires a flush to move a child to a newly added child element

### ISSUE THREAD

**Issue by serynth** [Polymer/polymer#1863]
This code snippet attempts to encapsulate the contents of the element's light DOM in a newly-created span.

If I do this ordinarily:
 
    var wrapper = document.createElement("span");
    Polymer.dom(this).appendChild(wrapper);
    Polymer.dom(wrapper).classList.add("wrapper");
    var nextChild;
    while ((nextChild = Polymer.dom(this).firstChild) !== Polymer.dom(this).lastChild) {
        Polymer.dom(wrapper).appendChild(nextChild);
    }

The contents of the element are merely `<span class="wrapper"></span>`

But if I flush after adding the span to the element:

    var wrapper = document.createElement("span");
    Polymer.dom(this).appendChild(wrapper);
    Polymer.dom(wrapper).classList.add("wrapper");
    Polymer.dom.flush();
    var nextChild;
    while ((nextChild = Polymer.dom(this).firstChild) !== Polymer.dom(this).lastChild) {
        Polymer.dom(wrapper).appendChild(nextChild);
    }

Then the span contains the appropriate contents.

Incidentally, moving the `Polymer.dom(this).appendChild(wrapper)` to the end of the while loop and changing the condition appropriately:

    var wrapper = document.createElement("span");
    Polymer.dom(wrapper).classList.add("wrapper");
    var nextChild;
    while ((nextChild = Polymer.dom(this).firstChild) !== null) {
        Polymer.dom(wrapper).appendChild(nextChild);
    }
    Polymer.dom(this).appendChild(wrapper);

Causes an error when appending the element's original children to the new wrapper element:
```
Uncaught TypeError: Cannot read property '_lightParent' of undefined
Polymer.DomApi.DomApi._removeNodeFromHost @ polymer-mini.html:587
Polymer.DomApi.DomApi.appendChild @ polymer-mini.html:460
```

Switching Polymer to use full Shadow DOM causes all four of these cases to work.

Oh, and also... When in Shady DOM, not using the Polymer DOM wrapper until the last appendChild:

    var wrapper = document.createElement("span");
    wrapper.classList.add("wrapper");
    var nextChild;
    while ((nextChild = Polymer.dom(this).firstChild) !== null) {
        wrapper.appendChild(nextChild);
    }
    Polymer.dom(this).appendChild(wrapper);

... sent my CPU usage for that tab into the stratosphere and made it completely, unrecoverably unresponsive. Incidentally, this also killed the JS Bin I was creating for this, so that's where the tests end. Sorry. :)

**Comment by kevinpschaaf** [comment]
We can't reproduce your first issue, so we didn't go through all of them.  Here's the your first code (called from `attached`), which seems to work fine: http://jsbin.com/zimixe/edit?html,output  Could you let us know how your use case differs (e.g., where are you calling that code from)?

Also note that modifying elements in your light DOM like this is generally considered an anti-pattern, since your light-DOM children aren't in your scope of control and below to you (the element's) owner.  Instead, a better use of this would be to use `<content></content>`, which is the Shadow DOM's approach to decorating elements passed into your light DOM, e.g. something like this: http://jsbin.com/wakoko/edit?html,output

In the event that you were doing this so that the `wrapper` class could be styled by the user, this would be an appropriate place to employ Polymer's [custom style shim](https://www.polymer-project.org/1.0/docs/devguide/styling.html#xscope-styling-details) to expose the wrapper styles as a custom property set:
http://jsbin.com/wakoko/edit?html,output

**Comment by serynth** [comment]
Testing for a bit, it seems like the difference that mattered was having a template for the element (even one that was just `<content></content>`), and it doesn't matter if it's done from inside or outside the element: http://jsbin.com/xoqewuwaxu/3/edit?html,output

It felt pretty bad, I'm not surprised to hear that it's an antipattern. ;) For what it's worth, I was trying to make an element that animates revealing the text within one character at a time, including any styling, in the style of a visual novel. It achieved this by wrapping the whole contents, then moving nodes outside one at a time (recursing on their contents if needed). The reason for the wrapper (with visibility: hidden) is because this prevents the text from visibly reflowing as it animates. A partially animated example might be

`This is some <b>bo<span class="invisible">ld</span></b><span class="invisible"> text!</span>`

I eventually decided that I was working against Polymer more than I was working with it and switched back away, although I may ask on StackOverflow what the better way is just for my own education for the next time I come back to Polymer. My guess is one of "clone the light DOM into the shadow DOM and then mess with it to your heart's content" or "translate the text from a string attribute into shadow DOM elements yourself and then mess with those to your heart's content".

**Comment by kevinpschaaf** [comment]
Ok, looks like that particular issue related to `<content>` has been resolved as of the latest release (v1.0.5): http://jsbin.com/lakoqe/edit?html,output

I'll go ahead and close the issue, but please feel free to inquire on slack or SO if you need help with approaches like that.
